controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp

```diff
@@ -0,0 +1,397 @@
+
+        FSM_MOVE_STATE(WAIT_FOR_SCAN_TRIGGERED);
+        break;
+    }
+    case eState::WAIT_FOR_SCAN_TRIGGERED: {
+        TASK_LOG(ERROR) << "This should not happen!";
+
+        m_last_scan_error_code = beerocks::DCS_SCAN_SCAN_TRIGGERED_EVENT_TIMEOUT;
+
+        FSM_MOVE_STATE(ABORT_SCAN);
+        break;
+    }
+    case eState::WAIT_FOR_RESULTS_READY: {
+        TASK_LOG(ERROR) << "This should not happen!";
+
+        m_last_scan_error_code = beerocks::DCS_SCAN_RESULTS_READY_EVENT_TIMEOUT;
+
+        FSM_MOVE_STATE(ABORT_SCAN);
+        break;
+    }
+    case eState::WAIT_FOR_RESULTS_DUMP: {
+        TASK_LOG(ERROR) << "This should not happen!";
+
+        m_last_scan_error_code = beerocks::DCS_SCAN_RESULTS_DUMP_EVENT_TIMEOUT;
+
+        FSM_MOVE_STATE(ABORT_SCAN);
+        break;
+    }
+    case eState::SCAN_DONE: {
+        LOG(TRACE) << "SCAN_DONE";
+
+        //update next continuous scan time
+        if (!m_is_single_scan) {
+            auto interval = std::chrono::seconds(database.get_dcs_interval_sec(m_radio_mac));
+            m_next_scan_timestamp_interval = m_last_scan_try_timestamp + interval;
+        }
+
+        m_last_scan_error_code = beerocks::DCS_SCAN_NO_ERROR;
+        database.set_dcs_last_scan_success(m_radio_mac, m_last_scan_error_code, m_is_single_scan);
+
+        database.set_dcs_scan_in_progress(m_radio_mac, false, m_is_single_scan);
+        m_is_single_scan = false;
+        FSM_MOVE_STATE(IDLE);
+        break;
+    }
+    case eState::ABORT_SCAN: {
+        LOG(ERROR) << "aborting scan for mac=" << m_radio_mac << ", last_scan_timestamp is not set";
+
+        database.set_dcs_last_scan_success(m_radio_mac, m_last_scan_error_code, m_is_single_scan);
+        database.set_dcs_scan_in_progress(m_radio_mac, false, m_is_single_scan);
+        m_is_single_scan = false;
+
+        FSM_MOVE_STATE(IDLE);
+        work(); // Skip pause, move to Idle immediately
+        break;
+    }
+    default: {
+        break;
+    }
+    }
+}
+
+void dynamic_channel_selection_task::handle_event(int event_type, void *obj)
+{
+    if (obj == nullptr) {
+        TASK_LOG(ERROR) << "obj == nullptr";
+        return;
+    }
+    auto event_radio_mac = network_utils::mac_to_string(((sMacAddr *)obj)->oct);
+    if (m_radio_mac == event_radio_mac) {
+        bool handle_event = false;
+        switch (eEvent(event_type)) {
+        case eEvent::TRIGGER_SINGLE_SCAN: {
+            TASK_LOG(DEBUG) << "TRIGGER_SINGLE_SCAN received";
+            auto trigger_single_scan_event = reinterpret_cast<sTriggerSingleScan_event *>(obj);
+            handle_event                   = true;
+            TASK_LOG(DEBUG) << "TRIGGER_SINGLE_SCAN handled on:"
+                            << network_utils::mac_to_string(
+                                   trigger_single_scan_event->radio_mac.oct);
+            m_is_single_scan_pending = true;
+            break;
+        }
+        case eEvent::SCAN_TRIGGERED: {
```

Comment:
We could, but I think that having the different events will help with the readability of the feature, since they are different event that have different meanings to them